rtrlib/aspa/aspa_private.h

```diff
@@ -41,82 +114,122 @@ struct aspa_store_node {
 	struct aspa_record record;
+	bool is_no_op;
 };
 
 /**
- * @brief ASPA update finalization arguments
- * @param unused_provider_arrays Array of pointers to now unused provider sets.
- * @param unused_provider_array_len Number of unused provider sets.
+ * @brief Computed ASPA update.
  */
-struct aspa_update_finalization_args {
-	uint32_t **unused_provider_arrays;
-	size_t unused_provider_array_len;
-#ifndef ASPA_UPDATE_IN_PLACE
-	struct aspa_array **old_array;
+struct aspa_update {
+	struct aspa_table *table;
+	struct aspa_update_operation *operations;
+	size_t operation_count;
+	struct aspa_update_operation *failed_operation;
+	struct aspa_store_node *node;
 	struct aspa_array *new_array;
-#endif
 };
 
 /**
- * @brief Updates the given ASPA table by adding and removing records.
+ * @brief Computes an update structure that can later be applied to the given ASPA table.
+ *
+ * @note Each record in an 'add' operation may have a provider array associated with it. Any record in a 'remove'
+ * operation must have its @c provider_count set to 0 and @c provider_array set to @c NULL .
+ * @note You must call @c aspa_table_finish_update afterwards.
  *
  * @param[in] aspa_table ASPA table to store new ASPA data in.
  * @param[in] rtr_socket The socket the updates originate from.
  * @param[in] operations  Add and remove operations to perform.
- * @param[in] len  Number of operations
- * @param[in] revert A boolean value indicating whether to apply the inverse the effect of the given operations (add <-> remove).
- * @param[in,out] failed_operation The operation responsible for causing the table update failure. @c NULL , if the update succeeded. If this function is called with @c revert set to @c false , treat this as an output parameter. If @c revert is @c true , this argument must not be @ NULL.
- * @param[out] cleanup_args Arguments for cleanup.
+ * @param[in] count  Number of operations.
+ * @param update The computed update. The update pointer must be non-NULL, but may point to a @c NULL value initially. Points to an update struct after  this function returns.
  * @return @c ASPA_SUCCESS On success.
  * @return @c ASPA_RECORD_NOT_FOUND If a records is supposed to be removed but cannot be found.
- * @return @c ASPA_DUPLICATE_RECORD If a records is supposed to be added but its corresponding customer ASN already exists.
+ * @return @c ASPA_DUPLICATE_RECORD If a records is supposed to be added but its corresponding customer ASN already
+ * exists.
  * @return @c ASPA_ERROR On on failure.
+ */
+enum aspa_status aspa_table_compute_update(struct aspa_table *aspa_table, struct rtr_socket *rtr_socket,
+					   struct aspa_update_operation *operations, size_t count,
+					   struct aspa_update **update);
+
+/**
+ * @brief Applys the given update, as previously computed by @c aspa_table_compute_update
+ * @param update The update that will be applied.
+ */
+void aspa_table_apply_update(struct aspa_update *update);
+
+/**
+ * @brief Finishes the update.
```

Comment:
provide more detail/explanation.
```suggestion
 * @brief Finishes the update: clears memory and unlocks @c update->table->update_lock .
```